stride/physics/iso_elastic/devito.py

```diff
@@ -0,0 +1,398 @@
+
+        else:
+            # If the source/receiver size has changed, then create new functions for them
+            if num_sources != self.dev_grid.vars.src.npoint:
+                self.dev_grid.sparse_time_function('src', num=num_sources, cached=False)
+
+            if num_receivers != self.dev_grid.vars.rec_tau.npoint:
+                self.dev_grid.sparse_time_function('rec', num=num_receivers, cached=False)
+
+        # Clear all buffers
+        self.dev_grid.vars.src.data_with_halo.fill(0.)
+        self.dev_grid.vars.rec_tau.data_with_halo.fill(0.)
+        self.dev_grid.vars.vel[0].data_with_halo.fill(0.)
+        self.dev_grid.vars.vel[1].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[0, 0].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[0, 1].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[1, 0].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[1, 1].data_with_halo.fill(0.)
+        self.boundary.clear()
+
+        # Set medium parameters
+        vp_with_halo = self.dev_grid.with_halo(vp.extended_data)
+        vs_with_halo = self.dev_grid.with_halo(vs.extended_data)
+        rho_with_halo = self.dev_grid.with_halo(rho.extended_data)
+
+        lam_with_halo = rho_with_halo * (vp_with_halo ** 2 - 2. * vs_with_halo ** 2)
+        mu_with_halo = self.dev_grid.with_halo(rho_with_halo * vs_with_halo ** 2)
+        byn_with_halo = self.dev_grid.with_halo(1 / rho_with_halo)
+
+        self.dev_grid.vars.lam_fun.data_with_halo[:] = lam_with_halo
+        self.dev_grid.vars.mu_fun.data_with_halo[:] = mu_with_halo
+        self.dev_grid.vars.byn_fun.data_with_halo[:] = byn_with_halo
+
+        # Set geometry and wavelet
+        wavelets = wavelets.data
+
+        self._src_scale = 1000
+        self._max_wavelet = np.max(np.abs(wavelets)) + 1e-31
+
+        window = scipy.signal.get_window(('tukey', 0.01), self.time.num, False)
+        window = window.reshape((self.time.num, 1))
+
+        self.dev_grid.vars.src.data[:] = wavelets.T * self._src_scale / self._max_wavelet * window
+
+        if self.interpolation_type == 'linear':
+            self.dev_grid.vars.src.coordinates.data[:] = shot.source_coordinates
+            self.dev_grid.vars.rec_tau.coordinates.data[:] = shot.receiver_coordinates
+
+    async def run_forward(self, wavelets, vp, vs, rho, **kwargs):
+        """
+        Run the state or forward problem.
+
+        Parameters
+        ----------
+        wavelets : Traces
+            Source wavelets.
+        vp : ScalarField
+            Compressional (acoustic) speed of sound of the medium, in [m/s].
+        vs : ScalarField
+            Transverse (shear) speed of sound of the medium, in [m/s].
+        rho : ScalarField
+            Density of the medium in [kg/m^3].
+        problem : Problem
+            Sub-problem being solved by the PDE.
+
+        Returns
+        -------
+
+        """
+        functions = dict(src=self.dev_grid.vars.src, )
```

Comment:
this dictionary should also include the receivers, see:

https://github.com/trustimaging/stride/blob/de3bcf1bea6ccd7125a753c1728b51fcadd43db0/stride/physics/iso_acoustic/devito.py#L325

Reply:
Correction made